### ISSUE TITLE
Shouldn't decode return an &mut T rather than an &T?

### ISSUE THREAD

**Issue by HadrienG2** [TimelyDataflow/abomonation#19]
As far as I can see, there is no reason why users shouldn't be allowed to modify the deserialized object in place if they so desire.

**Comment by frankmcsherry** [comment]
For some types this is ok, but for other types like `Vec<T>` the ability to invoke `&mut` method could prompt e.g. de-allocation of memory that the allocator is not expecting (in this case, halfway through some bytes array). You would also gain access to `mem::swap` and other methods that replace the serialized data with owned data, and would complicate the state of the world once the `&mut T` is dropped (a bunch of owned data behind that should be cleaned up).

Does that make sense / were you thinking of a different reason?

**Comment by HadrienG2** [comment]
Oh, I didn't realize that abomonation replaces the memory allocations of RAII objects with inline storage in its buffer. But in hindsight, it makes total sense, given how the `exhume` API looks.

A first question that comes to my mind is how you manage to get the pointer alignment right, but I guess this is already the topic of #8.

Also, this means that one must be extremely careful about implementing `Abomonation` for types that use internal mutability. It would probably be a good idea to mention this in the `Safety` section of the `Abomonation` trait documentation.

**Comment by frankmcsherry** [comment]
Yup. Some of these things have quieted a bit as it became (years back) clear that Rust didn't have a stable story around unsafe, and I didn't want to make specific claims until they sorted that out more. Mostly, I didn't want to give the false impression that watching out for a few footguns would mean safety ensues. Instead, the front-page warning of "do not use this on data you care about" is meant to cover that. Also, text like 

> Be warned that implementing `Abomonable` for types can be a giant disaster and is entirely discouraged.

though it looks like I have the wrong name there.

I'll add some text to the trait documentation after "do not call these methods" and before "maybe don't use the trait at all" warnings. :)

**Comment by frankmcsherry** [comment]
Btw, are you *using* this for something, or considering it? The crate exists mostly for a small-ish set of performance sensitive people working with big data, and it has other issues that limit broad applicability (e.g. no `HashMap` implementation, some other key owned library types). It's something I would love to make a bit more serious, but am blocked on the Rust unsafe team for a while, I think.

**Comment by HadrienG2** [comment]
I'm currently designing a `log` backend for real-time applications, and investigating the use of `abomonation` as a maximally-lightweight serialization mechanism when [transferring log records from the real-time threads to the non-real-time logger thread](https://github.com/rust-lang-nursery/log/issues/356). It seems that it could fit this sort of need really well.

But obviously, before actually using it, I had to go beyond your "this library will eat your data" blanket statements and get a reasonably solid understanding of what are abomonation's actual safety invariants. And this is how I ended up studying your API and spamming your bug tracker in the process ;)

By the way, you may be interested in the [unsafe-code-guidelines](https://github.com/rust-lang/unsafe-code-guidelines/) repo if you don't know about it already. There were a lot of interesting discussions in there, and the compiler people in there are always available to answer questions about how rustc currently behaves and whether they see a reasonable chance of it changing in the future.

**Comment by HadrienG2** [comment]
I think you should take a look at [this RustConf talk by @oli-obk](https://youtu.be/wkXNm_qo8aY). The notion of ConstRefSafe that they define for the purpose of allowing heap allocations in const-eval seems suspiciously to the type contract that you need for replacing heap allocations with inline allocations in abomonation...
